dv/verilator/memutil/cpp/verilator_memutil.cc

```diff
@@ -271,48 +275,59 @@ bool VerilatorMemUtil::ElfFileToBinary(const std::string &filepath,
     goto return_elf_end;
   }
 
-  GElf_Phdr phdr;
-  Elf_Data *elf_data;
-  elf_data = NULL;
-  for (size_t i = 0; i < phnum; i++) {
+  /*
+   * To mimic what objcopy does (that is, the binary target of BFD), we need to
+   * iterate over all loadable program headers, find the lowest address, and
+   * then copy in our loadable sections based on their offset with respect to
+   * the found base address.
+   */
+  for (i = 0; i < phnum; i++) {
     if (gelf_getphdr(elf_desc, i, &phdr) == NULL) {
       std::cerr << elf_errmsg(-1) << " segment number: " << i
                 << " in: " << filepath << std::endl;
       retval = false;
       goto return_elf_end;
     }
+
     if (phdr.p_type != PT_LOAD) {
-      std::cout << "Program header number " << i << "is not of type PT_LOAD."
-                << "Continue." << std::endl;
+      std::cout << "Program header number " << i << " is not of type PT_LOAD; "
+                << "ignoring." << std::endl;
       continue;
     }
+
+    if (phdr.p_filesz == 0)
+      continue;
+
+    if (!any || phdr.p_paddr < low)
+      low = phdr.p_paddr;
+
+    if (!any || phdr.p_paddr + phdr.p_filesz > high)
+      high = phdr.p_paddr + phdr.p_filesz;
+
+    any = true;
+  }
+
+  buf = (uint8_t *)malloc(len_bytes = (high - low));
```

Comment:
- can you please put the `len_bytes` assignment into a separate line?
- Can you add an `assert(buf)` after this statement?

Reply:
Fixed